### ISSUE TITLE
Set margins when hiding bar

### ISSUE THREAD

**Issue by Barbarossa93** [qtile/qtile#2378]
# Issue description
When issuing `qtile cmd-obj -o cmd -f hide_show_bar -a top` to hide the bar, the margins that were set in `config.py` are not applied ([this particular line](https://github.com/Barbarossa93/Genome/blob/ac3034b08f818334066750622c96593af5a6243e/.config/qtile/config.py#L1072)). When showing the bar after issuing the command again, the margins are reapplied (presumably because of https://github.com/qtile/qtile/commit/9305013bf487f20ba6b16ef6f2d1ebd1c60767db). Ideally, the margins that were set should stay there regardless of whether the bar is there or not.

Screenshot before hiding bar (and after showing it)
![1618154714](https://user-images.githubusercontent.com/57952969/114310458-d00b6600-9ab8-11eb-9049-444dfeec5483.png)

Screenshot after hiding bar
![1618154736](https://user-images.githubusercontent.com/57952969/114310474-d863a100-9ab8-11eb-9c12-9aeddf6892a2.png)

# Qtile version

0.17.1.dev142+ge0b752c7

# Stack traces

I think this is irrelevant, but let me know if you need this.

# Configuration

https://github.com/Barbarossa93/Genome/blob/main/.config/qtile/config.py

**Comment by tych0** [comment]
I'm confused; you want to leave empty space for the bar when it is hidden?

**Comment by Barbarossa93** [comment]
Not quite; there's an option to specify the `margins` for the bar (see [this line](https://github.com/Barbarossa93/Genome/blob/ac3034b08f818334066750622c96593af5a6243e/.config/qtile/config.py#L1072)). I'm saying that the margin should still be respected when the bar is hidden. In the screenshot, the top part only has a margin of 9px (the margin defined by the layout) while the gaps around the edges are all 18px. The defined margin for below the bar is set to 21px. When I issue that command to hide the bar, the 21px margin is ignored. It is however reapplied when I show the bar again.

**Comment by Barbarossa93** [comment]
I sincerely don't mean to offend, but this is the relevant part of the docs http://docs.qtile.org/en/latest/manual/ref/commands.html?highlight=Bar#bar if this helps refresh your memory

**Comment by tych0** [comment]
> I'm saying that the margin should still be respected when the bar is hidden.

Right, this is what I don't understand :). I see the visual artifact in your screenshot, but that feels like "working as intended" to me.

If you want to write a patch to enable this behavior that's fine, but let's keep the default behavior the way it is.

**Comment by Barbarossa93** [comment]
I guess when I think of the bar being hidden (as is implied by `hide_show_bar`), I think of strictly the bar being hidden, not including the margins around the bar. I think I see your point in that the margins are defined as part of the bar and so they should be treated as the bar itself as well, thus working as intended. What I could do alternatively is to maybe define a top gap when the bar is hidden, but a Gap is considered to be a bar; so wouldn't that overwrite the bar that is already defined?

**Comment by tych0** [comment]
Yeah, I don't think we support multiple bars (there are some closed bugs about it), but implementing that might be another route to accomplish this.

**Comment by ramnes** [comment]
I implemented multiple gaps per position in #1387 but the IPC interface wasn't good enough. Feel free to revive the PR. :)

**Comment by Barbarossa93** [comment]
Unfortunately I don't have the time to implement this myself. I'm still trying to figure out how to do https://github.com/qtile/qtile/issues/2115. Maybe someday I can get around to this 😭

**Comment by m-col** [comment]
You might be able to get the behaviour you want by swapping the bar with a gap at `screen.top` and laying out the screen again, kinda like what happens in `hide_show_bar` except there is a gap present. TBH I'd use this too, the reason I never hide my bar is because I don't like that lack of a gap :p if that's sufficient, it might be simpler to allow passing a `Gap` to `hide_show_bar`, or even just an `int` to create a `Gap` on the fly, which might be more cli friendly.

**Comment by LamprosPitsillos** [comment]
I have a problem where when i hide the bar and bring it back again the bottom margin is ignored.. Am asking around  if anyone has had the same problem before i open an issue.

**Comment by zackfall** [comment]
@m-col is there a chance that what you said get implemented in the future?
